File: leetcode/problem_23.py
### Problem 23. Merge k Sorted Lists (Hard): https://leetcode.com/problems/merge-k-sorted-lists/
### Tags: Linked List, Heap

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Time Complexity: O(NlogK) where N is the total number of elements in all the linked lists and K is the number of linked lists
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        result = None
        for i in range(len(lists)):
            head = lists[i]
            while(head):
                heapq.heappush(min_heap, head.val)
                head = head.next
            print(result)
        tail = None
        while(min_heap):
            popped = heapq.heappop(min_heap)
            if(result is None):
                result = ListNode(popped)
                tail = result
            else:
                tail.next = ListNode(popped)
                tail = tail.next
        return result

# Time Complexity: O(NlogN) where N is the total number of elements in all the linked lists
class Solution2:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        result = []
        for i in range(len(lists)):
            head = lists[i]
            while(head):
                result.append(head.val)
                head = head.next
        arr = sorted(result)
        tail = None
        result = None
        i = 0
        while(i < len(arr)):
            popped = arr[i]
            if(result is None):
                result = ListNode(popped)
                tail = result
            else:
                tail.next = ListNode(popped)
                tail = tail.next
            i += 1
        return result

File: leetcode/problem_23.kt
/**
 * Problem 23. Merge k Sorted Lists (Hard)
 * tags: Array, Linked List, Heap
 */
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

 // time complexity: O(nlogk), n is the number of elements in the list
// space complexity: O(k) , k is the number of elements in the list
class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        /**
        *   input: List<Node>
        *   output: head:List<Node>
        *   Example : [[2, 5, 7], [1, 2, 3], [3, 4, 7]]
        *           pq = [[1, 2 ,3], [2, 5, 7], [3, 4, 7]]
        *           pq = [[2 ,3], [2, 5, 7], [3, 4, 7]]
        *           
        *                   tail: [2]
        *                   head: [1, 2]
        *           
        */
        var priorityQ = PriorityQueue<ListNode>(compareBy {it.`val`})
        var head: ListNode? = null
        var tail: ListNode? = null
        for(node in lists){ // time complexity: O(k)
            if(node != null){
                priorityQ.add(node) // time complexity: O(logk)
            }
        }

        while(priorityQ.size > 0){  // time complexity for this loop: O(n)
            var pointer = priorityQ.poll()  // time complexity: O(logk)
            val temp = ListNode(pointer.`val`)
            if(head == null){
                head = temp
                tail = temp
            }else{
                tail?.next = temp
                tail = tail?.next
            }
            pointer = pointer.next
            if(pointer != null){
                priorityQ.add(pointer)  // time complexity: O(logk)
            }
        }
        return head
    }
}